README.md
<p align="center">
  <img height="150" src="https://d1wzvcwrgjaybe.cloudfront.net/repos/manwaring/lambda-wrapper/readme-category-icon.png">
  <img height="150" src="https://d1wzvcwrgjaybe.cloudfront.net/repos/manwaring/lambda-wrapper/readme-repo-icon.png">
</p>

<p align="center">
  <a href="https://npmjs.com/package/@manwaring/lambda-wrapper">
    <img src="https://flat.badgen.net/npm/v/@manwaring/lambda-wrapper?icon=npm&label=@latest"></a>
  <a href="https://www.npmjs.com/package/@manwaring/lambda-wrapper">
    <img src="https://flat.badgen.net/npm/dt/@manwaring/lambda-wrapper?icon=npm"></a>
  <a href="https://codecov.io/gh/manwaring/lambda-wrapper">
    <img src="https://flat.badgen.net/codecov/c/github/manwaring/lambda-wrapper/?icon=codecov"></a>
  <a href="https://packagephobia.now.sh/result?p=@manwaring/lambda-wrapper">
    <img src="https://flat.badgen.net/packagephobia/install/@manwaring/lambda-wrapper?icon=packagephobia"></a>
</p>

<p align="center">
  <a href="https://circleci.com/gh/manwaring/lambda-wrapper">
    <img src="https://flat.badgen.net/circleci/github/manwaring/lambda-wrapper/master?icon=circleci"></a>
  <a href="https://flat.badgen.net/dependabot/manwaring/lambda-wrapper">
    <img src="https://flat.badgen.net/dependabot/manwaring/lambda-wrapper/?icon=dependabot&label=dependabot"></a>
  <a href="https://www.npmjs.com/package/@manwaring/lambda-wrapper">
    <img src="https://flat.badgen.net/github/license/manwaring/lambda-wrapper"></a>
</p>

# AWS Lambda wrapper library

### This documentation is for v4 of the library - [go here](old-docs) for [v1](old-docs/v1/README.md), [v2](old-docs/v2/README.md), and [v3](old-docs/v3/README.md) documentation.

1. [Overview](#overview)
1. [Installation and setup](#installation-and-setup)
1. [Supported events](#supported-events)
   - [API Gateway](#api-gateway)
   - [API Gateway HTTP API](#api-gateway-http-api)
   - [CloudFormation Custom Resource](#cloudformation-custom-resource)
   - [DynamoDB Stream](#dynamodb-stream)
   - [Lambda Authorizer](#lambda-authorizer)
   - [SNS](#sns)
   - [Generic event](#generic-event)
1. [Example projects](#example-projects)

# Overview

### TL;DR

This library provides custom Lambda function wrappers which expose standard, abstracted functionality so that developers can focus on writing business logic instead of parsing event payloads and crafting response objects.

### Rationale and motivation

AWS Lambda supports a wide variety of event triggers, each with unique payloads and expected response objects. The Lambda method signature, however, only provides a raw event object and has no included mechanisms for simplifying payload parsing or response object creation. For example API Gateway events include only the raw request body, leaving it up to developers to implement parsing themselves. Similarly the developer is responsible for creating a response object which includes the correct HTTP status code and headers. This library exposes helpful abstractions like parsed HTTP bodies based on content-type headers, and success functions which create response objects with the correct status codes and headers for returning to API Gateway.

_Feedback is appreciated! If you have an idea for how this library can be improved (or just a complaint/criticism) then [please open an issue](https://github.com/manwaring/lambda-wrapper/issues/new)._

# Installation and setup

Install and save the package:

`npm i -S @manwaring/lambda-wrapper`

`yarn add @manwaring/lambda-wrapper`

## Optional configuration

If you want the wrapper to log request and response messages (helpful for debugging) set an environemnt variable for `LAMBDA_WRAPPER_LOG=true`.

If you want each invocation to be tagged with the AWS region, stage/environment, and Git revision simply set environment variables for each and the library will pick them up, for example `REGION=us-east-1`, `STAGE=prod`, `REVISION=f4ba682`. See [git-rev-sync](https://www.npmjs.com/package/git-rev-sync) and [serverless-plugin-git-variables](https://www.npmjs.com/package/serverless-plugin-git-variables) for libraries that can help you set git revision automatically.

# Supported events

Each event listed here has a wrapper which provides a deconstructable method signature exposing parsed/unmarshalled request parameters and helper response methods.

1. [API Gateway](#api-gateway)
1. [API Gateway HTTP API](#api-gateway-http-api)
1. [CloudFormation Custom Resource](#cloudformation-custom-resource)
1. [DynamoDB Stream](#dynamodb-stream)
1. [Lambda Authorizer](#lambda-authorizer)
1. [SNS](#sns)
1. [Generic event](#generic-event) (a basic wrapper with support for success and failure responses)

# API Gateway

## Sample TypeScript implementation

```ts
import { api } from '@manwaring/lambda-wrapper';
import { CustomInterface } from './custom-interface';
import { doSomething } from './you-code';

export const handler = api<CustomInterface>(async ({ body, path, success, invalid, error }) => {
  try {
    const { pathParam1, pathParam2 } = path;
    if (!pathParam1) {
      return invalid();
    }
    const results = await doSomething(body, pathParam1, pathParam2);
    return success({ body: results });
  } catch (err) {
    return error({ err });
  }
});
```

By passing in CustomInterface as a generic type the method signature will cast the `body` object as an instance of CustomInterface, making TypeScript development easier.  Note that the type is not required and the body property defaults to type `any`.

<details>
<summary>Sample implementation without generic</summary>

```ts
import { api } from '@manwaring/lambda-wrapper';
import { doSomething } from './you-code';

export const handler = api(async ({ body, path, success, invalid, error }) => {
  try {
    const { pathParam1, pathParam2 } = path;
    if (!pathParam1) {
      return invalid();
    }
    const results = await doSomething(body, pathParam1, pathParam2);
    return success({ body: results });
  } catch (err) {
    return error({ err });
  }
});
```
</details>

## Properties and methods available on wrapper signature

<details open>
<summary>Deconstructable wrapper signature</summary>

Note that all properties are undefined if not present on the original request.

```ts
export interface ApiSignature<T = any> {
  event: APIGatewayEvent; // original event provided by AWS
  body: T; // body payload parsed according to content-type headers (or raw if no content-type headers found) and cast as T if provided (defaults to `any`)
  websocket: WebsocketRequest; // websocket connection payload
  path: { [name: string]: string }; // path params as key-value pairs
  query: { [name: string]: string }; // query params as key-value pairs
  headers: { [name: string]: string }; // headers as key-value pairs
  testRequest: boolean; // indicates if this is a test request - looks for a header matching process.env.TEST_REQUEST_HEADER (dynamic from application) or 'Test-Request' (default)
  auth: any; // auth context from custom authorizer
  success(params?: ResponseParameters): ApiResponse;
  invalid(params?: ResponseParameters): ApiResponse;
  notFound(params?: ResponseParameters): ApiResponse;
  notAuthorized(params?: ResponseParameters): ApiResponse;
  redirect(params: RedirectParameters): ApiResponse;
  error(params?: ErrorParameters): ApiResponse;
  custom(params: CustomParameters): ApiResponse;
}
```

</details>

<details>
<summary>ApiResponse</summary>

```ts
interface ApiResponse {
  statusCode: number;
  headers: { [name: string]: any };
  body?: string;
}
```

</details>

<details>
<summary>ResponseParameters</summary>

```ts
interface ResponseParameters {
  body?: any; // response body
  json?: boolean; // indicates if body should be JSON-stringified and content-type header set to application/json, defaults to true
  cors?: boolean; // indicates if CORS headers should be added, defaults to true
  statusCode?: number; // status code to return, defaults by callback (success: 200, invalid: 400, notFound: 404, notAuthorized: 401, redirect: 302, error: 500)
  headers?: { [key: string]: any }; // custom headers to include
}
```

</details>

<details>
<summary>RedirectParameters</summary>

```ts
interface RedirectParameters {
  url: string; // url to redirect to
  cors?: boolean; // indicates if CORS headers should be added, defaults to true
  statusCode?: number; // status code to return, defaults to 302
  headers?: { [key: string]: any }; // custom headers to include
}
```

</details>

<details>
<summary>ErrorParameters</summary>

```ts
interface ErrorParameters {
  body?: any; // response body
  json?: boolean; // indicates if body should be JSON-stringified and content-type header set to application/json, defaults to true
  cors?: boolean; // indicates if CORS headers should be added, defaults to true
  statusCode?: number; // status code to return, defaults to 500
  headers?: { [key: string]: any }; // custom headers to include
  err?: Error; // optional Error object for automatic logging
}
```

</details>

<details>
<summary>CustomParameters</summary>

```ts
interface CustomParameters {
  body?: any; // response body
  json?: boolean; // indicates if body should be JSON-stringified and content-type header set to application/json, defaults to true
  cors?: boolean; // indicates if CORS headers should be added, defaults to true
  statusCode: number; // status code to return
  headers?: { [key: string]: any }; // custom headers to include
  err?: Error; // optional Error object for automatic logging
}
```

</details>

<details>
<summary>WebsocketRequest</summary>

```ts
export interface WebsocketRequest {
  accountId: string;
  apiId: string;
  connectedAt?: number;
  connectionId?: string;
  domainName?: string;
  domainPrefix?: string;
  eventType?: string;
  extendedRequestId?: string;
  protocol: string;
  httpMethod: string;
  identity: APIGatewayEventIdentity;
  messageDirection?: string;
  messageId?: string | null;
  path: string;
  stage: string;
  requestId: string;
  requestTime?: string;
  requestTimeEpoch: number;
  resourceId: string;
  resourcePath: string;
  routeKey?: string;
}
```

</details>

## Response functions

<details open>
<summary>Success</summary>

### Available parameters

```ts
{
  body?: any,
  json?: boolean,
  cors?: boolean,
  statusCode?: number,
  headers?: { [key: string]: any}
}
```

### Default parameters

```ts
{
  json: true,
  cors: true,
  statusCode: 200
}
```

### Invocation with defaults

```ts
const response = { hello: 'world' };
return success({ body: response });

// returns
{
  body: "{\"hello\":\"world\"}",
  statusCode: 200,
  headers: {
    'Access-Control-Allow-Origin': '*',
    'Access-Control-Allow-Credentials': true,
    'Content-Type': 'application/json'
  }
}
```

### Invocation overriding defaults

```ts
const response = '<svg xmlns="http://www.w3.org/2000/svg"></svg>';
const headers = { 'Content-Type': 'image/svg+xml' };
return success({ body: response, json: false, cors: false, headers });

// returns
{
  body: "<svg xmlns=\"http://www.w3.org/2000/svg\"></svg>",
  statusCode: 200,
  headers: { 'Content-Type': 'image/svg+xml' }
}
```

</details>

<details>
<summary>Invalid</summary>

### Available parameters

```ts
{
  body?: any,
  json?: boolean,
  cors?: boolean,
  statusCode?: number,
  headers?: { [key: string]: any}
}
```

### Default parameters

```ts
{
  json: true,
  cors: true,
  statusCode: 400
}
```

### Invocation with defaults

```ts
return invalid();

// returns
{
  statusCode: 400,
  headers: {
    'Access-Control-Allow-Origin': '*',
    'Access-Control-Allow-Credentials': true,
  }
}
```

### Invocation overriding defaults

```ts
const response = { invalid: 'properties' };
return invalid({ body: response, cors: false });

// returns
{
  body: "{\"invalid\":\"properties\"}",
  statusCode: 400,
  headers: { 'Content-Type': 'application/json' }
}
```

</details>

<details>
<summary>Not found</summary>

### Available parameters

```ts
{
  body?: any,
  json?: boolean,
  cors?: boolean,
  statusCode?: number,
  headers?: { [key: string]: any}
}
```

### Default parameters

```ts
{
  json: true,
  cors: true,
  statusCode: 404
}
```

### Invocation with defaults

```ts
return notFound();

// returns
{
  statusCode: 404,
  headers: {
    'Access-Control-Allow-Origin': '*',
    'Access-Control-Allow-Credentials': true,
  }
}
```

### Invocation overriding defaults

```ts
const response = 'Not found';
return notFound({ body: response, cors: false });

// returns
{
  body: "Not found",
  statusCode: 404,
}
```

</details>

<details>
<summary>Not authorized</summary>

### Available parameters

```ts
{
  body?: any,
  json?: boolean,
  cors?: boolean,
  statusCode?: number,
  headers?: { [key: string]: any}
}
```

### Default parameters

```ts
{
  json: true,
  cors: true,
  statusCode: 401
}
```

### Invocation with defaults

```ts
return notAuthorized();

// returns
{
  statusCode: 401,
  headers: {
    'Access-Control-Allow-Origin': '*',
    'Access-Control-Allow-Credentials': true,
  }
}
```

### Invocation overriding defaults

```ts
const response = 'Not authorized';
return notAuthorized({ body: response, cors: false });

// returns
{
  body: "Not Authorized",
  statusCode: 401,
}
```

</details>

<details>
<summary>Redirect</summary>

### Available parameters

```ts
{
  url: string,
  cors?: boolean,
  statusCode?: number,
  headers?: { [key: string]: any}
}
```

### Default parameters

```ts
{
  cors: true,
  statusCode: 302
}
```

### Invocation with defaults

```ts
const url = 'https://github.com/manwaring/lambda-wrapper';
return redirect({ url });

// returns
{
  statusCode: 302,
  headers: {
    'Access-Control-Allow-Origin': '*',
    'Access-Control-Allow-Credentials': true,
    'Location': 'https://github.com/manwaring/lambda-wrapper'
  }
}
```

### Invocation overriding defaults

```ts
const url = 'https://github.com/manwaring/lambda-wrapper';
return redirect({ url, statusCode: 308, cors: false });

// returns
{
  statusCode: 308,
  headers: {
    'Location': 'https://github.com/manwaring/lambda-wrapper'
  }
}
```

</details>

<details>
<summary>Error</summary>

### Available parameters

```ts
{
  body?: any,
  json?: boolean,
  cors?: boolean,
  statusCode?: number,
  headers?: { [key: string]: any},
  err?: Error
}
```

### Default parameters

```ts
{
  json: true,
  cors: true,
  statusCode: 500
}
```

### Invocation with defaults

```ts
return error();

// returns
{
  statusCode: 500,
  headers: {
    'Access-Control-Allow-Origin': '*',
    'Access-Control-Allow-Credentials': true,
  }
}
```

### Invocation overriding defaults

```ts
catch (err) {
  const body = { error: 'Unexpected error' };
  return error({ body, err });
}

// logs
console.debug(err);

// returns
{
  body: "{\"error\": \"Unexpected error\"}",
  statusCode: 500,
  headers: {
    'Access-Control-Allow-Origin': '*',
    'Access-Control-Allow-Credentials': true,
  }
}
```

</details>

<details>
<summary>Custom</summary>

### Available parameters

```ts
{
  body?: any,
  json?: boolean,
  cors?: boolean,
  statusCode: number,
  headers?: { [key: string]: any},
  err?: Error
}
```

### Default parameters

```ts
{
  json: true,
  cors: true,
}
```

### Invocation with defaults

```ts
return custom({ statusCode: 418 });

// returns
{
  statusCode: 418,
  headers: {
    'Access-Control-Allow-Origin': '*',
    'Access-Control-Allow-Credentials': true,
  }
}
```

### Invocation overriding defaults

```ts
const body = { message: 'Custom response' };
return custom({ body, statusCode: 418 });

// returns
{
  body: "{\"message\": \"Custom response\"}",
  statusCode: 418,
  headers: {
    'Access-Control-Allow-Origin': '*',
    'Access-Control-Allow-Credentials': true,
  }
}
```

</details>

# API Gateway HTTP API

Other than the raw payload from AWS the HTTP API method signature and response functions match the API Gateway signature and functions. Hooray for wrappers! Note that you still need to provide the correct wrapper function so that the library can parse the AWS event correctly.

## Sample TypeScript implementation

```ts
import { httpApi } from '@manwaring/lambda-wrapper';
import { CustomInterface } from './custom-interface';
import { doSomething } from './you-code';

export const handler = httpApi<CustomInterface>(async ({ body, path, success, invalid, error }) => {
  try {
    const { pathParam1, pathParam2 } = path;
    if (!pathParam1) {
      return invalid();
    }
    const results = await doSomething(body, pathParam1, pathParam2);
    return success({ body: results });
  } catch (err) {
    return error({ err });
  }
});
```

By passing in CustomInterface as a generic type the method signature will cast the `body` object as an instance of CustomInterface, making TypeScript development easier.  Note that the type is not required and the body property defaults to type `any`.

<details>
<summary>Sample implementation without generic</summary>

```ts
import { httpApi } from '@manwaring/lambda-wrapper';
import { doSomething } from './you-code';

export const handler = httpApi(async ({ body, path, success, invalid, error }) => {
  try {
    const { pathParam1, pathParam2 } = path;
    if (!pathParam1) {
      return invalid();
    }
    const results = await doSomething(body, pathParam1, pathParam2);
    return success({ body: results });
  } catch (err) {
    return error({ err });
  }
});
```
</details>

## Properties and methods available on wrapper signature

<details open>
<summary>Deconstructable wrapper signature</summary>

Note that all properties are undefined if not present on the original request.

```ts
export interface HttpApiSignature<T = any> {
  event: HttpApiEvent; // original event provided by AWS
  body: T; // body payload parsed according to content-type headers (or raw if no content-type headers found) and cast as T if provided (defaults to `any`)
  rawPath: string; // the endpoint path used to invoke this Lambda
  path: { [name: string]: string }; // path params as key-value pairs
  query: { [name: string]: string }; // query params as key-value pairs
  rawQueryString: string // the raw query string from the request
  headers: { [name: string]: string }; // headers as key-value pairs
  testRequest: boolean; // indicates if this is a test request - looks for a header matching process.env.TEST_REQUEST_HEADER (dynamic from application) or 'Test-Request' (default)
  auth: any; // auth context from JWT authorizer
  success(params?: ResponseParameters): ApiResponse;
  invalid(params?: ResponseParameters): ApiResponse;
  notFound(params?: ResponseParameters): ApiResponse;
  notAuthorized(params?: ResponseParameters): ApiResponse;
  redirect(params: RedirectParameters): ApiResponse;
  error(params?: ErrorParameters): ApiResponse;
  custom(params: CustomParameters): ApiResponse;
}
```

</details>

<details>
<summary>ApiResponse</summary>

```ts
interface ApiResponse {
  statusCode: number;
  headers: { [name: string]: any };
  body?: string;
}
```

</details>

<details>
<summary>ResponseParameters</summary>

```ts
interface ResponseParameters {
  body?: any; // response body
  json?: boolean; // indicates if body should be JSON-stringified and content-type header set to application/json, defaults to true
  cors?: boolean; // indicates if CORS headers should be added, defaults to true
  statusCode?: number; // status code to return, defaults by callback (success: 200, invalid: 400, notFound: 404, notAuthorized: 401, redirect: 302, error: 500)
  headers?: { [key: string]: any }; // custom headers to include
}
```

</details>

<details>
<summary>RedirectParameters</summary>

```ts
interface RedirectParameters {
  url: string; // url to redirect to
  cors?: boolean; // indicates if CORS headers should be added, defaults to true
  statusCode?: number; // status code to return, defaults to 302
  headers?: { [key: string]: any }; // custom headers to include
}
```

</details>

<details>
<summary>ErrorParameters</summary>

```ts
interface ErrorParameters {
  body?: any; // response body
  json?: boolean; // indicates if body should be JSON-stringified and content-type header set to application/json, defaults to true
  cors?: boolean; // indicates if CORS headers should be added, defaults to true
  statusCode?: number; // status code to return, defaults to 500
  headers?: { [key: string]: any }; // custom headers to include
  err?: Error; // optional Error object for automatic logging
}
```

</details>

<details>
<summary>CustomParameters</summary>

```ts
interface CustomParameters {
  body?: any; // response body
  json?: boolean; // indicates if body should be JSON-stringified and content-type header set to application/json, defaults to true
  cors?: boolean; // indicates if CORS headers should be added, defaults to true
  statusCode: number; // status code to return
  headers?: { [key: string]: any }; // custom headers to include
  err?: Error; // optional Error object for automatic logging
}
```

</details>

<details>
<summary>HttpApiEvent</summary>

_[AWS documentation of raw event](https://docs.aws.amazon.com/apigateway/latest/developerguide/http-api-develop-integrations-lambda.html#http-api-develop-integrations-lambda.proxy-format)_

```ts
export interface HttpApiEvent {
  version: string;
  routeKey: string;
  rawPath: string;
  rawQueryString: string;
  cookies: string[];
  headers: { [key: string]: string };
  queryStringParameters: { [key: string]: string };
  requestContext: {
    accountId: string;
    apiId: string;
    authorizer: {
      jwt: {
        claims: { [key: string]: string };
        scopes: string[];
      };
    };
    domainName: string;
    domainPrefix: string;
    http: {
      method: string;
      path: string;
      protocol: string;
      sourceIp: string;
      userAgent: string;
    };
    requestId: string;
    routeKey: string;
    stage: string;
    time: string;
    timeEpoch: number;
  };
  body: string;
  pathParameters: { [key: string]: string };
  isBase64Encoded: boolean;
  stageVariables: { [key: string]: string };
}
```

</details>

## Response functions

<details open>
<summary>Success</summary>

### Available parameters

```ts
{
  body?: any,
  json?: boolean,
  cors?: boolean,
  statusCode?: number,
  headers?: { [key: string]: any}
}
```

### Default parameters

```ts
{
  json: true,
  cors: true,
  statusCode: 200
}
```

### Invocation with defaults

```ts
const response = { hello: 'world' };
return success({ body: response });

// returns
{
  body: "{\"hello\":\"world\"}",
  statusCode: 200,
  headers: {
    'Access-Control-Allow-Origin': '*',
    'Access-Control-Allow-Credentials': true,
    'Content-Type': 'application/json'
  }
}
```

### Invocation overriding defaults

```ts
const response = '<svg xmlns="http://www.w3.org/2000/svg"></svg>';
const headers = { 'Content-Type': 'image/svg+xml' };
return success({ body: response, json: false, cors: false, headers });

// returns
{
  body: "<svg xmlns=\"http://www.w3.org/2000/svg\"></svg>",
  statusCode: 200,
  headers: { 'Content-Type': 'image/svg+xml' }
}
```

</details>

<details>
<summary>Invalid</summary>

### Available parameters

```ts
{
  body?: any,
  json?: boolean,
  cors?: boolean,
  statusCode?: number,
  headers?: { [key: string]: any}
}
```

### Default parameters

```ts
{
  json: true,
  cors: true,
  statusCode: 400
}
```

### Invocation with defaults

```ts
return invalid();

// returns
{
  statusCode: 400,
  headers: {
    'Access-Control-Allow-Origin': '*',
    'Access-Control-Allow-Credentials': true,
  }
}
```

### Invocation overriding defaults

```ts
const response = { invalid: 'properties' };
return invalid({ body: response, cors: false });

// returns
{
  body: "{\"invalid\":\"properties\"}",
  statusCode: 400,
  headers: { 'Content-Type': 'application/json' }
}
```

</details>

<details>
<summary>Not found</summary>

### Available parameters

```ts
{
  body?: any,
  json?: boolean,
  cors?: boolean,
  statusCode?: number,
  headers?: { [key: string]: any}
}
```

### Default parameters

```ts
{
  json: true,
  cors: true,
  statusCode: 404
}
```

### Invocation with defaults

```ts
return notFound();

// returns
{
  statusCode: 404,
  headers: {
    'Access-Control-Allow-Origin': '*',
    'Access-Control-Allow-Credentials': true,
  }
}
```

### Invocation overriding defaults

```ts
const response = 'Not found';
return notFound({ body: response, cors: false });

// returns
{
  body: "Not found",
  statusCode: 404,
}
```

</details>

<details>
<summary>Not authorized</summary>

### Available parameters

```ts
{
  body?: any,
  json?: boolean,
  cors?: boolean,
  statusCode?: number,
  headers?: { [key: string]: any}
}
```

### Default parameters

```ts
{
  json: true,
  cors: true,
  statusCode: 401
}
```

### Invocation with defaults

```ts
return notAuthorized();

// returns
{
  statusCode: 401,
  headers: {
    'Access-Control-Allow-Origin': '*',
    'Access-Control-Allow-Credentials': true,
  }
}
```

### Invocation overriding defaults

```ts
const response = 'Not authorized';
return notAuthorized({ body: response, cors: false });

// returns
{
  body: "Not Authorized",
  statusCode: 401,
}
```

</details>

<details>
<summary>Redirect</summary>

### Available parameters

```ts
{
  url: string,
  cors?: boolean,
  statusCode?: number,
  headers?: { [key: string]: any}
}
```

### Default parameters

```ts
{
  cors: true,
  statusCode: 302
}
```

### Invocation with defaults

```ts
const url = 'https://github.com/manwaring/lambda-wrapper';
return redirect({ url });

// returns
{
  statusCode: 302,
  headers: {
    'Access-Control-Allow-Origin': '*',
    'Access-Control-Allow-Credentials': true,
    'Location': 'https://github.com/manwaring/lambda-wrapper'
  }
}
```

### Invocation overriding defaults

```ts
const url = 'https://github.com/manwaring/lambda-wrapper';
return redirect({ url, statusCode: 308, cors: false });

// returns
{
  statusCode: 308,
  headers: {
    'Location': 'https://github.com/manwaring/lambda-wrapper'
  }
}
```

</details>

<details>
<summary>Error</summary>

### Available parameters

```ts
{
  body?: any,
  json?: boolean,
  cors?: boolean,
  statusCode?: number,
  headers?: { [key: string]: any},
  err?: Error;
}
```

### Default parameters

```ts
{
  json: true,
  cors: true,
  statusCode: 500
}
```

### Invocation with defaults

```ts
return error();

// returns
{
  statusCode: 500,
  headers: {
    'Access-Control-Allow-Origin': '*',
    'Access-Control-Allow-Credentials': true,
  }
}
```

### Invocation overriding defaults

```ts

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/), and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html)

## [4.4.3]&nbsp;&nbsp;(2021-04-10)

### Updated

- Updated dependencies

## [4.4.2]&nbsp;&nbsp;(2020-11-25)

### Fixed

- Update calculation of `testRequest` prop on api signature based on headers

## [4.4.1]&nbsp;&nbsp;(2020-11-25)

### Changed

- Set `'Access-Control-Allow-Headers': 'Test-Request'` for support of the test-request header when `cors=true` is specified in API responses

## [4.4.0]&nbsp;&nbsp;(2020-11-24)

### Changed

- Added `'Access-Control-Allow-Headers': '*'` when `cors=true` is specified in API responses
- Update dependencies

## [4.3.0]&nbsp;&nbsp;(2020-11-13)

### Added

- Added rawQueryString as parsed param to http api wrapper

## [4.2.0]&nbsp;&nbsp;(2020-11-13)

### Added

- Added rawPath as parsed param to http api wrapper

### Changed

- Update dependencies

## [4.1.1]&nbsp;&nbsp;(2020-10-31)

### Changed

- Update documentation

## [4.1.0]&nbsp;&nbsp;(2020-10-31)

### Added

- Custom response option for api gateway responses

### Changed

- Update dependencies

## [4.0.6]&nbsp;&nbsp;(2020-09-23)

### Changed

- Update dependencies

## [4.0.5]&nbsp;&nbsp;(2020-09-11)

### Changed

- Update dependencies

## [4.0.4]&nbsp;&nbsp;(2020-09-06)

### Changed

- Better debug logging of events - include original and parsed messages instead of just parsed

## [4.0.3]&nbsp;&nbsp;(2020-09-06)

### Changed

- Update dependencies

## [4.0.2]&nbsp;&nbsp;(2020-08-16)

### Changed

- Update dependencies
- Update documentation
- Update example project

## [4.0.1]&nbsp;&nbsp;(2020-08-11)

### Fixed

- Updated api response interfaces to indicate optional vs required params

## [4.0.0]&nbsp;&nbsp;(2020-08-11)

### Changed

- Updated the API Gateway and HTTP API response function signatures for easier overriding of defaults

## [3.7.0]&nbsp;&nbsp;(2020-08-03)

### Added

- Support for HTTP API 2.0 payloads

## [3.6.3]&nbsp;&nbsp;(2020-06-21)

### Fixed

- Fix changes to eagerly parse api payloads as JSON

## [3.6.2]&nbsp;&nbsp;(2020-06-21)

### Updated

- Try harder to parse api payloads as JSON even if no content-type headers are available

## [3.6.1]&nbsp;&nbsp;(2020-06-21)

### Updated

- Log websocket payload from api wrapper when log levels set to info

## [3.6.0]&nbsp;&nbsp;(2020-06-21)

### Added

- Add support for Websocket connections

### Updated

- Updated all libraries to latest

## [3.5.0]&nbsp;&nbsp;(2020-05-10)

### Changed

- API response 'not authorized' now returns 401 instead of 403

## [3.4.0]&nbsp;&nbsp;(2020-05-10)

### Added

- Added support for the API Gateway v2 HTTP API Auth property

## [3.3.3]&nbsp;&nbsp;(2020-05-08)

### Changed

- Change npmignore config to whitelisting

## [3.3.2]&nbsp;&nbsp;(2020-05-07)

### Changed

- Make generics optional by defaulting to any

## [3.3.1]&nbsp;&nbsp;(2020-05-07)

### Changed

- Update dependencies

## [3.3.0]&nbsp;&nbsp;(2020-05-07)

### Changed

- Update API, SNS, and generic wrappers with support for generic types

## [3.2.0]&nbsp;&nbsp;(2020-05-02)

### Changed

- DynamoDB stream wrapper accepts optional type generic for better TypeScript support
- DynamoDB stream wrapper returns string enum for DynamoDB event type instead of 'string'

## [3.1.3]&nbsp;&nbsp;(2020-03-16)

### Changed

- Updated out of date libraries

## [3.1.2]&nbsp;&nbsp;(2020-02-11)

### Fixed

- Support additional properties on content-type
- Add correct content-type header for JSON stringified API responses
- Update old libraries

## [3.1.1]&nbsp;&nbsp;(2020-01-22)

### Added

- Add a 'not authorized' response handler to the api wrapper

## [3.1.0]&nbsp;&nbsp;(2020-01-20)

### Changed

- Default signature params to undefined instead of null to allow for default values
- Updated libraries

## [3.0.0]&nbsp;&nbsp;(2020-01-05)

### Changed

- Api error response generates a 500 status code instead of 503

## [2.1.1]&nbsp;&nbsp;(2020-01-05)

### Added

- Api error response logs Error message if passed an Error object

## [2.1.0]&nbsp;&nbsp;(2020-01-04)

### Changed

- Api error response includes Error message if passed an Error object

### Fixed

- Removed extra files from published package
- Correct package.json entry point to lib

## [2.0.1]&nbsp;&nbsp;(2020-01-04)

### Fixed

- Removed extra files from published package
- Correct package.json entry point to lib

## [2.0.0]&nbsp;&nbsp;(2020-01-04)

### Changed

- Removed references to Epsagon and IOPipe
- Instead of callbacks, response methods use the recommended return syntax

### Added

- V2 of the docs

## [1.2.2]&nbsp;&nbsp;(2019-10-14)

### Changed

- Modified console out logging level when serverless-specific framework libraries are not found

## [1.2.1]&nbsp;&nbsp;(2019-10-06)

### Fixed

- Added correct type definitions for new API success response method

## [1.2.0]&nbsp;&nbsp;(2019-10-06)

### Added

- Support for JSON.stringify replacer in the API success response method

## [1.1.4]&nbsp;&nbsp;(2019-09-10)

### Added

- Added missing wrapper documentation

## [1.1.3]&nbsp;&nbsp;(2019-09-10)

### Changed

- Reverted CloudFormation Custom Resource responder library - realized that issue was with async function instead of lib

## [1.1.2]&nbsp;&nbsp;(2019-09-09)

### Changed

- Updated CloudFormation Custom Resource responder library

## [1.1.1]&nbsp;&nbsp;(2019-09-08)

### Changed

- Updated CloudFormation Custom Resource responder library

## [1.1.0]&nbsp;&nbsp;(2019-09-08)

### Changed

- Switch to a new CloudFormation Custom Resource responder library

## [1.0.10]&nbsp;&nbsp;(2019-09-08)

### Added

- Example projects using the latest version of the library

### Changed

- Clean up documentation

## [1.0.9]&nbsp;&nbsp;(2019-09-07)

### Changed

- Eliminate duplicative logging from response functions

## [1.0.8]&nbsp;&nbsp;(2019-09-07)


src/api/api.ts
import { APIGatewayEvent, Context, Callback } from 'aws-lambda';
import { Metrics, Body } from '../common';

const metrics = new Metrics('API Gateway');

const HEADERS = {
  'Access-Control-Allow-Origin': '*',
  'Access-Control-Allow-Credentials': true
};

export function apiWrapper<T extends Function>(fn: T): T {
  return <any>function(event: APIGatewayEvent, context: Context, callback: Callback) {
    const { body, path, query, request, auth, headers, testRequest } = getRequestFields(event);
    metrics.common(request);

    function success(payload: any = null): void {
      const response = { statusCode: 200, headers: HEADERS };
      if (payload) {
        response['body'] = JSON.stringify(payload);
      }
      metrics.success(payload);
      return callback(null, response);
    }

    function invalid(errors: string[] = []): void {
      const response = { statusCode: 400, headers: HEADERS, body: JSON.stringify({ errors, request }) };
      metrics.invalid(response);
      return callback(null, response);
    }

    function redirect(url: string): void {
      HEADERS['Location'] = url;
      const response = { statusCode: 302, headers: HEADERS };
      metrics.redirect(response);
      return callback(null, response);
    }

    function error(error: any = ''): void {
      metrics.error(error);
      return callback(error);
    }

    const signature: ApiSignature = {
      event,
      body,
      path,
      query,
      headers,
      testRequest,
      auth,
      success,
      invalid,
      redirect,
      error
    };
    return fn(signature);
  };
}

function getRequestFields(event: APIGatewayEvent): any {
  const path = event.pathParameters ? event.pathParameters : null;
  const query = event.queryStringParameters ? event.queryStringParameters : null;
  const auth = event.requestContext && event.requestContext.authorizer ? event.requestContext.authorizer : null;
  const headers = event.headers ? event.headers : null;
  const body = new Body(event.body, headers).getParsedBody();
  const TEST_REQUEST_HEADER = process.env.TEST_REQUEST_HEADER || 'Test-Request';
  const testRequest = headers && headers[TEST_REQUEST_HEADER] ? JSON.parse(headers[TEST_REQUEST_HEADER]) : false;
  const request = { body, path, query, auth, headers, testRequest };
  return { body, path, query, auth, request, headers, testRequest };
}

export interface ApiSignature {
  event: APIGatewayEvent; // original event
  body: any; // JSON parsed body payload if exists (otherwise null)
  path: { [name: string]: string }; // path param payload as key-value pairs if exists (otherwise null)
  query: { [name: string]: string }; // query param payload as key-value pairs if exists (otherwise null)
  headers: { [name: string]: string }; // header payload as key-value pairs if exists (otherwise null)
  testRequest: boolean; // indicates if this is a test request - looks for a header matching process.env.TEST_REQUEST_HEADER (dynamic from application) or 'Test-Request' (default)
  auth: any; // auth context from custom authorizer if exists (otherwise null)
  success(payload?: any): void; // returns 200 status with payload
  invalid(errors?: string[]): void; // returns 400 status with errors in payload
  redirect(url: string): void; // returns 302 redirect with new url
  error(error?: any): void; // returns 500 status with error and original request payload
}
